kythe/go/extractors/config/runextractor/cmakecmd/cmakecmd.go

```diff
@@ -14,43 +14,203 @@
+	fs.StringVar(&c.buildDir, "builddir", "", "An optional path to the directory in which to build. If empty, defaults to a unique subdirectory of sourcedir.")
+}
+
+func (c *cmakeCommand) verifyFlags() error {
+	for _, key := range []string{"KYTHE_CORPUS", "KYTHE_ROOT_DIRECTORY", "KYTHE_OUTPUT_DIRECTORY"} {
+		if os.Getenv(key) == "" {
+			return fmt.Errorf("required %s not set", key)
+		}
+	}
+	if c.extractor == "" {
+		return fmt.Errorf("required -extractor not set")
+	}
+	if c.sourceDir == "" {
+		return fmt.Errorf("required -sourcedir not set")
+	}
+	return nil
 }
 
-// Execute implements the subcommands interface and runs cmake extraction.
 func (c *cmakeCommand) Execute(ctx context.Context, fs *flag.FlagSet, args ...interface{}) subcommands.ExitStatus {
-	// if err := exec.Command("run", "some", "thing") {
-	//   c.Fail("some error %v", err)
+	if err := c.verifyFlags(); err != nil {
+		return c.Fail("incorrect flags: %v", err)
+	}
+	// Since we have to change our working directory, resolve all of our paths early.
+	extractor, err := filepath.Abs(c.extractor)
+	if err != nil {
+		return c.Fail("unable to resolve path to extractor: %v", err)
+	}
+	sourceDir, err := filepath.Abs(c.sourceDir)
+	if err != nil {
+		return c.Fail("unable to resolve source directory: %v", err)
+	}
+	var buildDir string
+	if c.buildDir == "" {
+		// sourceDir is already an absolute directory
+		buildDir = filepath.Join(sourceDir, "build-"+uuid.New())
+		if err := os.Mkdir(buildDir, 0755); err != nil {
+			// Unlike below, we need to fail if the "unique" directory exists.
+			return c.Fail("unable to create build directory: %v", err)
+		}
+		// Only clean up the build directory if it was unspecified.
+		defer cleanBuild(buildDir)
+	} else {
+		buildDir, err = filepath.Abs(c.buildDir)
+		if err != nil {
+			return c.Fail("unable to resolve build directory: %v", err)
+		}
+	}
+
+	// Create the build directory if it doesn't already exist.
+	if err := os.MkdirAll(buildDir, 0755); err != nil {
+		return c.Fail("unable to create build directory: %v", err)
+	}
+
+	if err := runIn(exec.CommandContext(ctx, "cmake", "-DCMAKE_EXPORT_COMPILE_COMMANDS=ON", sourceDir), buildDir); err != nil {
+		return c.Fail("error configuring cmake: %v", err)
+	}
+
+	if err := runIn(exec.CommandContext(ctx, "cmake", "--build", "."), buildDir); err != nil {
+		return c.Fail("error building repository: %v", err)
+	}
+
+	// TODO(shahms): Move compile_commands.json handling to common library.
+	commands, err := readCommands(filepath.Join(buildDir, "compile_commands.json"))
+	if err != nil {
+		return c.Fail("unable to read compile_commands.json: %v", err)
+	}
+	exEnv, err := extractorEnv(buildDir)
+	if err != nil {
+		return c.Fail("unable to set up environment: %v", err)
+	}
+	sem := semaphore.NewWeighted(128) // Limit concurrency.
+	extraction, ctx := errgroup.WithContext(ctx)
```

Comment:
I suggest pulling this out into a function, though of course it works fine as-is.

Reply:
Was in the middle of that when you responded; done.